File: solutions/101.symmetric-tree/symmetric-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
const isSymmetric = root => {
  if (!root) return true

  return compare(root.left, root.right)
}

const compare = (left, right) => {
  if (!left || !right) {
    return left === right
  }

  if (left.val !== right.val) return false
  return compare(left.left, right.right) && compare(left.right, right.left)
}

File: solutions/101.symmetric-tree/symmetric-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def isSymmetric(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if not root:
            return True

        def check(left, right):
            if not left and not right:
                return True

            if left and right:
                if left.val != right.val:
                    return False
                return check(left.left, right.right) and check(left.right, right.left)

            return False

        return check(root.left, root.right)
